Audio and sound support

Taeir opened the pull request:
Added audio and sound support.

I will add the tests later, otherwise the pull request would become even larger.

Overview of what class does what:
```
AudioEngine:	Singleton class for all things audio.

FishClip: 		Decorator for javax.sound.sampled.Clip, which adds
				functionality to check if a clip is done playing and to
				update the clip volume dynamically when the volume
				properties in AudioEngine change.

Sound:  		An object to represent a playable sound. It caches the
				data from an input stream into a byte array, and offers
				an easy method to get new FishClips of a sound.

ClipRunnable:	Runnable that plays a clip when ran.

VolumeListener: A functional interface for listening for volume changes.
```

Advised order of review:
- Changes to pom.xml + added dependencies.
- VolumeListener
- ClipRunnable
- FishClip
- Sound
- AudioEngine

<feature

Taeir commented:
Made small change to the layout.

<layout

MrCrime commented on src/main/java/com/github/fishio/audio/FishClip.java:
```diff
@@ -0,0 +1,113 @@
+public class FishClip implements LineListener, AutoCloseable {
+	private Clip clip;
+	private volatile boolean done;
+	private boolean effect;
```
I got confused what "effect" means until I read the JavaDoc in the constructor that is a SOUND effect. So I suggest renaming it to "soundEffect".

Taeir replied:
@MrCrime Good point, will do

<naming

MrCrime commented on src/main/java/com/github/fishio/audio/Sound.java:
```diff
@@ -0,0 +1,158 @@
+	private static final int READ_BUFFER_SIZE = 16384;
+	
+	private AudioFormat format;
+	private boolean effect;
```
See my previous comment

MrCrime commented on src/main/java/com/github/fishio/audio/Sound.java:
```diff
@@ -0,0 +1,158 @@
+			//We want the file in the following format, so we can play it.
+			this.format = new AudioFormat(AudioFormat.Encoding.PCM_SIGNED, 
+					ais.getFormat().getSampleRate(),
+                    16,
```
Is this just github, or is the layout a little bit messed up here? with the tabs/spaces.

<layout

Taeir replied:
@MrCrime It's just github, but I will see if there is a bunch of spaces in there or something.

<layout

jessearens commented on src/main/java/com/github/fishio/audio/Sound.java:
```diff
@@ -0,0 +1,158 @@
+	 */
+	public boolean isEffect() {
+		return effect;
+	}
```
Do you really need the same isEffect() method in both the Sound and FishClip classes? Shouldn't the effect boolean of the FishClip class be linked to the effect boolean of the Sound class, as the sound file defines whether the sound is background music or a sound effect? Perhaps I do not really understand the connection between Sound and FishClip correctly.

< organisation

jessearens commented on src/main/java/com/github/fishio/audio/FishClip.java:
```diff
@@ -0,0 +1,113 @@
+	public boolean isEffect() {
+		return effect;
+	}
+	
```
See my comment at the isEffect() method in the Sound class.

<organisation,review

jessearens commented:
I commented on one method which is duplicated in two classes. It's probably my misunderstanding, but if you could clarify that for me, it would be splendid. The rest seems good to me.

<feature

ChielBruin commented:
I can confirm it also works on linux!
Good job!

<feature

Taeir commented:
I implemented the name and layout changes as suggested by @MrCrime and removed a method as suggested by @jessearens .

<layout,naming

LMiljak commented:
You could consider making the loading of music/sound effects in a seperate class AudioLoader or something. It would make the AudioEngine class a little bit smaller and have less responsibilities.

<organisation

jessearens commented:
I agree with Luka on this. It fits the Responsibility-Driven-Design approach better than the current implementation. If you fix that, I think your pull request is good to go!

<organisation

Taeir commented:
@MrCrime @jessearens wouldn't this turn AudioLoader into a dataclass, or into a class with which AudioEngine has a lot of interaction, aka something that should not be separated?

<organisation

LMiljak commented:
@Taeir 

Not really, if you place two methods in the AudioLoader loadEffects and loadMusic and make them return a collection of Sound objects, then the AudioLoader is not storing anything that the AudioEngine is accessing. What could be stored in the AudioLoader are the file paths (Strings) of the sounds, but the AudioEngine won't have acces to those. Then the AudioLoader actually provides a service instead of just being a data class storing Sound objects.

<organisation

Diocruel commented on src/main/java/com/github/fishio/audio/AudioEngine.java:
```diff
@@ -0,0 +1,416 @@
+	/**
+	 * The maximum amount of threads that can play sounds concurrently.
+	 */
+	private static final int MAX_THREAD_COUNT = 11;
```
Why eleven?

Taeir replied:
See the requirements: "The audio engine shall support playing at least 10 sounds at the same time". Since the background music is always playing, 11 - 1 = 10.

<layout

Diocruel commented:
It looks really awesome! Did you remove the christmass steel drums yet? We really don't need that tune. I do suggest looking at the other comments too!

<review.

Taeir commented:
Made quite a few changes based on the comments:
- Implemented the AudioLoader class as suggested by @MrCrime.
- Added SoundEffects loading support, which searches the jar for .wav or .mp3 files in the sound/effects folder (or subfolders)
- Reworked loading music to be uniform with loading SoundEffects. This means that loading music in a random order is now possible.
- Music is now started when the program is opened.
- Fix for not stopping music when closing the game.
- Added button sound effect (but not yet implemented)

<feature,organisation,bug

Diocruel commented:
The new changes look good to me!
+1

<review
